File: backend/data/calculations.py
"""Calculo de metricas derivadas (espejo de lib/calculations.ts del frontend).

Se mantiene en el backend para poder servir analytics calculados en Python
(curva de TNA y estadisticas) y para validar la logica del lado servidor.
"""

from __future__ import annotations

import calendar
from datetime import date, datetime
from typing import Dict, List, Optional

DAY_COUNT = 365

_MONTHS = {
    "ENE": 1, "FEB": 2, "MAR": 3, "ABR": 4, "MAY": 5, "JUN": 6,
    "JUL": 7, "AGO": 8, "SEP": 9, "OCT": 10, "NOV": 11, "DIC": 12,
}


def month_expiry(label: str) -> str:
    """Convierte 'JUN26' -> ISO del ultimo dia del mes ('2026-06-30').

    Heuristica usada como fallback cuando el mercado no informa maturityDate.
    """
    code, year_suffix = label[:3].upper(), label[3:]
    month = _MONTHS.get(code)
    if month is None or not year_suffix.isdigit():
        return ""
    year = 2000 + int(year_suffix)
    last_day = calendar.monthrange(year, month)[1]
    return date(year, month, last_day).isoformat()


def days_to_expiry(expiry_iso: str, as_of: Optional[date] = None) -> int:
    """Dias calendario hasta el vencimiento (>= 0)."""
    if not expiry_iso:
        return 0
    as_of = as_of or date.today()
    expiry = datetime.fromisoformat(expiry_iso).date()
    return max(0, (expiry - as_of).days)


def implied_tna(futuro: Optional[float], spot: float, days: int) -> Optional[float]:
    """TNA implicita: ((Futuro / Spot) - 1) * (365 / Dias)."""
    if futuro is None or not spot or days <= 0:
        return None
    return (futuro / spot - 1) * (DAY_COUNT / days)


def implied_tea(futuro: Optional[float], spot: float, days: int) -> Optional[float]:
    """TEA implicita: (Futuro / Spot) ^ (365 / Dias) - 1."""
    if futuro is None or not spot or days <= 0:
        return None
    return (futuro / spot) ** (DAY_COUNT / days) - 1


def spread(bid: Optional[float], ask: Optional[float]) -> Optional[float]:
    """Spread = Ask - Bid."""
    if bid is None or ask is None:
        return None
    return ask - bid


def base(futuro: Optional[float], spot: float) -> Optional[float]:
    """Base / forward points = Futuro - Spot."""
    if futuro is None:
        return None
    return futuro - spot


def build_analytics(frame: Dict) -> Dict:
    """Calcula la curva de TNA y estadisticas agregadas a partir de un frame.

    Demuestra el calculo quant del lado servidor (consumible en /api/analytics).
    """
    spot = float(frame.get("spot", {}).get("value") or 0)
    curve: List[Dict] = []
    tnas: List[float] = []
    spreads: List[float] = []
    volumes: List[float] = []
    ois: List[float] = []

    for c in frame.get("contracts", []):
        mark = c.get("settlement") if c.get("settlement") is not None else c.get("last")
        days = days_to_expiry(c.get("expiry", ""))
        tna = implied_tna(mark, spot, days)
        if tna is not None:
            curve.append({"ticker": c["ticker"], "tna": tna, "days": days})
            tnas.append(tna)
        sp = spread(c.get("bid"), c.get("ask"))
        if sp is not None:
            spreads.append(sp)
        volumes.append(float(c.get("volume") or 0))
        ois.append(float(c.get("openInterest") or 0))

    def avg(values: List[float]) -> Optional[float]:
        return sum(values) / len(values) if values else None

    return {
        "curve": curve,
        "stats": {
            "totalVolume": sum(volumes),
            "totalOpenInterest": sum(ois),
            "avgVolume": avg(volumes),
            "avgOpenInterest": avg(ois),
            "avgSpread": avg(spreads),
            "maxTna": max(tnas) if tnas else None,
            "minTna": min(tnas) if tnas else None,
        },
    }

File: lib/calculations.ts
/**
 * Motor de calculo del monitor.
 *
 * Centraliza TODA la logica derivada (TNA, TEA, base, spread, forward points,
 * dias al vencimiento y estadisticas agregadas). Tanto el feed simulado como el
 * backend PyRofex producen `RawMarketFrame`; aqui se transforman en
 * `MarketSnapshot` + `MarketStats`. No hay duplicacion de formulas.
 */

import type {
  ComputedContract,
  FutureContract,
  MarketSnapshot,
  MarketStats,
  PriceDirection,
  RawMarketFrame,
} from "./types";

const MS_PER_DAY = 86_400_000;
export const DAY_COUNT = 365;

/** Dias calendario hasta el vencimiento (>= 0). */
export function daysToExpiry(expiryISO: string, from: number): number {
  const expiry = new Date(`${expiryISO}T00:00:00`).getTime();
  return Math.max(0, Math.round((expiry - from) / MS_PER_DAY));
}

/** TNA implicita: ((Futuro / Spot) - 1) * (365 / Dias). */
export function impliedTNA(
  futuro: number | null,
  spot: number,
  days: number,
): number | null {
  if (futuro == null || !spot || days <= 0) return null;
  return (futuro / spot - 1) * (DAY_COUNT / days);
}

/** TEA implicita: (Futuro / Spot) ^ (365 / Dias) - 1. */
export function impliedTEA(
  futuro: number | null,
  spot: number,
  days: number,
): number | null {
  if (futuro == null || !spot || days <= 0) return null;
  return Math.pow(futuro / spot, DAY_COUNT / days) - 1;
}

/** Spread = Ask - Bid. */
export function spread(bid: number | null, ask: number | null): number | null {
  if (bid == null || ask == null) return null;
  return ask - bid;
}

/** Base (forward points) = Futuro - Spot. */
export function base(futuro: number | null, spot: number): number | null {
  if (futuro == null) return null;
  return futuro - spot;
}

/** Precio de referencia para metricas: ajuste, sino ultimo. */
function markPrice(c: FutureContract): number | null {
  return c.settlement ?? c.last ?? null;
}

function direction(current: number | null, previous: number | null): PriceDirection {
  if (current == null || previous == null) return 0;
  if (current > previous) return 1;
  if (current < previous) return -1;
  return 0;
}

/** Calcula un contrato derivado a partir del crudo. */
export function computeContract(
  c: FutureContract,
  spot: number,
  now: number,
  prevLast: number | null,
): ComputedContract {
  const mark = markPrice(c);
  const days = daysToExpiry(c.expiry, now);
  const change =
    mark != null && c.referenceClose != null ? mark - c.referenceClose : null;
  const changePct =
    mark != null && c.referenceClose
      ? mark / c.referenceClose - 1
      : null;

  return {
    ...c,
    daysToExpiry: days,
    change,
    changePct,
    spread: spread(c.bid, c.ask),
    base: base(mark, spot),
    forwardPoints: base(mark, spot),
    tna: impliedTNA(mark, spot, days),
    tea: impliedTEA(mark, spot, days),
    oiChange:
      c.prevOpenInterest != null ? c.openInterest - c.prevOpenInterest : null,
    priceDir: direction(c.last, prevLast),
  };
}

/** Construye el snapshot completo + estadisticas desde un frame crudo. */
export function buildSnapshot(
  frame: RawMarketFrame,
  prev?: MarketSnapshot,
): { snapshot: MarketSnapshot; stats: MarketStats } {
  const prevLast = new Map<string, number | null>();
  prev?.contracts.forEach((c) => prevLast.set(c.ticker, c.last));

  const spotValue = frame.spot.value;
  const contracts = frame.contracts.map((c) =>
    computeContract(c, spotValue, frame.timestamp, prevLast.get(c.ticker) ?? null),
  );

  const snapshot: MarketSnapshot = {
    timestamp: frame.timestamp,
    a3500: frame.a3500,
    spot: frame.spot,
    contracts,
    source: frame.source,
  };

  return { snapshot, stats: computeStats(contracts) };
}

/** Estadisticas agregadas de la rueda. */
export function computeStats(contracts: ComputedContract[]): MarketStats {
  const n = contracts.length || 1;
  const totalVolume = sum(contracts.map((c) => c.volume));
  const totalOpenInterest = sum(contracts.map((c) => c.openInterest));
  const oiChanges = contracts
    .map((c) => c.oiChange)
    .filter((v): v is number => v != null);
  const spreads = contracts
    .map((c) => c.spread)
    .filter((v): v is number => v != null);
  const withTna = contracts.filter((c) => c.tna != null);

  return {
    totalVolume,
    totalOpenInterest,
    oiChangeTotal: oiChanges.length ? sum(oiChanges) : null,
    avgVolume: totalVolume / n,
    avgOpenInterest: totalOpenInterest / n,
    avgSpread: spreads.length ? sum(spreads) / spreads.length : null,
    mostLiquid: argmax(contracts, (c) => c.volume),
    frontContract: argmin(
      contracts.filter((c) => c.daysToExpiry >= 0),
      (c) => c.daysToExpiry,
    ),
    maxTna: argmax(withTna, (c) => c.tna ?? -Infinity),
    minTna: argmin(withTna, (c) => c.tna ?? Infinity),
  };
}

// ----------------------------- utilidades -----------------------------

function sum(values: number[]): number {
  return values.reduce((acc, v) => acc + v, 0);
}

function argmax<T>(items: T[], key: (item: T) => number): T | null {
  if (!items.length) return null;
  return items.reduce((best, item) => (key(item) > key(best) ? item : best));
}

function argmin<T>(items: T[], key: (item: T) => number): T | null {
  if (!items.length) return null;
  return items.reduce((best, item) => (key(item) < key(best) ? item : best));
}
